### ISSUE TITLE
With 4.x release behavior of decoding url parameters (path and query parameters) has changed.

### ISSUE THREAD

**Issue by mshambharkar** [aspnet/AspNetKatana#393]
I recently updated the version of Microsoft.Owin.dll from 3.0 to 4.1 and found a couple of issues that impact backward compatibility.
I have this url ``` http://localhost:${Port}/api/values/${Parameter}/${Parameter}?query1=${Parameter}&query2=${Parameter}&query3=${Parameter} ``` , where parameter contains url encoded characters (trying to pass URL reserved characters in the route as path and query parameter.
Simple Controller and it's action is defined as 
``` csharp
[RoutePrefix("api/values")]
public class ValuesController : ApiController
{
    [HttpGet]
    [Route("{Path1}/{Path2}")]
    public object Get([FromUri]string Path1, [FromUri] string Path2, string query1, string query2, string query3)
    {
        return new ResponseModel(Path1, Path2, query1, query2, query3);
    }
}
```
**In 3.x release, Owin (self-hosted Web API server running) would decode only once before passing those parameters to ActionFilter or Action. In 4.x parameters are decoded twice before passing it to ActionFilter or Action.**
For example:
Parameter = :?#[]@"!$&'()*,;=
| Owin version | Encoding level | Encoded parameter                                                                     | Path                                                | Query                                               |
|--------------|----------------|---------------------------------------------------------------------------------------|-----------------------------------------------------|-----------------------------------------------------|
| 3.x          | 1              | %3A%3F%23%5B%5D%40%22%21%24%26%27%28%29%2A%2C%3B%3D                                   | :?#[]@"!$&'()*,;=                                   | :?#[]@"!$&'()*,;=                                   |
| 4.x          | 1              |                                                                                       | :?#[]@"!$&'()*,;=                                   | :?#[]@"!$&'()*,;=                                   |
| 3.x          | 2              | %253A%253F%2523%255B%255D%2540%2522%2521%2524%2526%2527%2528%2529%252A%252C%253B%253D | %3A%3F%23%5B%5D%40%22%21%24%26%27%28%29%2A%2C%3B%3D | %3A%3F%23%5B%5D%40%22%21%24%26%27%28%29%2A%2C%3B%3D |
| 4.x          | 2              |                                                                                       | :?#[]@"!$&'()*,;=                                   | %3A%3F%23%5B%5D%40%22%21%24%26%27%28%29%2A%2C%3B%3D                                   |

The above table is just a sample and there are other possible scenarios like adding '+' & '/' to the parameter string and encoding three times at the client side before making a request.

I see a couple of issues

1.  Breaking change from 3.x to 4.x, the client now needs to encode 3 times the same string which was working with 2 times encoding. 3 times encoding is required if you want to pass  '/' as a parameter.
2.  Inconsistent decoding at Path parameters and Query parameters, path parameters are decoded twice whereas query params keep older behavior of 3.x.

After looking at a code and after several trials, I believe this behavior in 4.x was introduced with a resolution of bug [104](https://github.com/aspnet/AspNetKatana/issues/104)

I have placed sample server and test client code at GitHub repo for reference (it also includes other scenarios to verify) [AspNetKatana-Issue](https://github.com/mshambharkar/AspNetKatana-Issue)

**Comment by Tratcher** [comment]
#104 (#135) affected encoding, not decoding. Can you explain more why you think it's connected here?

**Comment by mshambharkar** [comment]
I agree code is related to encoding and not decoding and it is just my understanding, would need to debug (owin and webapi assemblies both) to ascertain the issue.
There are multiple reasons for my thinking
1. I am using Web API self-hosted configuration, I am only modifying 3 libraries - Microsoft.owin.dll,Microsoft.Owin.Hosting.dll and Microsoft.Owin.Host.HttpListener.dll keeping other libraries at a fixed version (verified behavior with Owin 3.0 and 4.0 ) 
2. Comparing code of Microsoft.Owin/PathString.cs between two versions, 3.0 code would always encode the input (even encoded string as well) whereas 4.0 code would encode only when unescaped characters are present. This behavior is consistent with Path parameters behavior and their decoding.
3. In 4.0 release only Bug #104 was related to Uri 

I believe WebApi pipeline is also taking care of decoding at 1 level and that is the reason the difference in this encoding of the path is creating a difference at Action and ActionFilter.

**Comment by minune-msft** [comment]
We are observing a similar behavior updating from version 3.0.1 to 4.1.1.

Previously are sending requests that can look like this, (note the percentage symbol):
/object/Object-:%25_*?a689c3ce-90a3-4317-a260-dfc61ad3e1f6

The object value is extracted using the [FromUri] attribute

On version 3.0.1, the [FromUri] attribute resolves to: Device-:%25_*?a689c3ce-90a3-4317-a260-dfc61ad3e1f6

On version 4.1.1 it resolves to Device-:%_*?a689c3ce-90a3-4317-a260-dfc61ad3e1f6

**Comment by ificator** [comment]
I'm seeing a similar issue upgrading from `3.0.1` to `4.2.2` and it looks like #135 is the root cause. The change to `ToUriComponent()` purported to be a backport of https://github.com/aspnet/HttpAbstractions/commit/150b4708f158c0cdcc2d49f2ec72f3e82c6c64e4 , however it went further and added a check for `isPercentEncodedChar`. If such a character is detect it does NOT encode the `%`, which makes the output of `ToUriComponent()` fundamentally wrong.

Since `ToUriComponent()` is used in `PathString.ToString()`, which is in turn used by `OwinRequest.Uri` it must represent an encoded string. Now if the caller provides `/test%2520file.txt` the value of `request.Uri` is `/test%20file.txt` which is not correct.
